File: packages/contracts/python/src/cuvoy_contracts/preferences.py
"""Trip input, extracted preferences, and Trip Controls (PROJECT_SPEC §4, §13, §18)."""

from datetime import date

from pydantic import Field, model_validator

from cuvoy_contracts.common import ContractModel
from cuvoy_contracts.constants import (
    DEFAULT_DAY_END_LOCAL,
    DEFAULT_DAY_START_LOCAL,
    DEFAULT_DINNER_END,
    DEFAULT_DINNER_START,
    DEFAULT_LUNCH_END,
    DEFAULT_LUNCH_START,
    MEAL_MIN_DURATION_MINUTES,
)
from cuvoy_contracts.enums import (
    BudgetTier,
    GroupPriority,
    LocationType,
    MaxTransitPreset,
    OwnedVehicle,
    Pace,
    PublicTransportMode,
)


class BudgetInput(ContractModel):
    daily_amount: float = Field(..., gt=0)
    currency: str = Field(..., min_length=1)
    raw: str | None = None
    tier: BudgetTier | None = None


class TravelDates(ContractModel):
    start_date: date | None = None
    end_date: date | None = None
    duration_days: int | None = Field(default=None, ge=1)

    @model_validator(mode="after")
    def require_span(self) -> "TravelDates":
        has_range = self.start_date is not None and self.end_date is not None
        if not has_range and self.duration_days is None:
            raise ValueError("Provide start_date+end_date or duration_days")
        if has_range and self.end_date < self.start_date:
            raise ValueError("end_date must be on or after start_date")
        return self


class LocationInput(ContractModel):
    query: str = Field(..., min_length=1)
    type: LocationType | None = None
    radius_km: float | None = Field(default=None, gt=0)


class TransportationPreference(ContractModel):
    owns_vehicle: bool
    vehicle: OwnedVehicle | None = None
    public_mode: PublicTransportMode | None = None

    @model_validator(mode="after")
    def resolve_mode(self) -> "TransportationPreference":
        if self.owns_vehicle:
            if self.vehicle is None:
                raise ValueError("vehicle is required when owns_vehicle is true")
        elif self.public_mode is None:
            self.public_mode = PublicTransportMode.MIXED
        return self


class Traveler(ContractModel):
    name: str | None = None
    interests: list[str] = Field(default_factory=list)
    is_team_lead: bool = False


class GroupPlanning(ContractModel):
    enabled: bool = False
    travelers: list[Traveler] = Field(default_factory=list)
    priority: GroupPriority = GroupPriority.EVERYONE


class AccessibilityPreferences(ContractModel):
    kids: bool = False
    elderly: bool = False
    wheelchair: bool = False
    notes: str | None = None


class FoodPreferences(ContractModel):
    dietary_restrictions: list[str] = Field(default_factory=list)
    cuisines: list[str] = Field(default_factory=list)


class MealWindow(ContractModel):
    start_local: str
    end_local: str
    min_duration_minutes: int = Field(default=MEAL_MIN_DURATION_MINUTES, ge=60)


class TripControls(ContractModel):
    max_transit_preset: MaxTransitPreset = MaxTransitPreset.BALANCED
    max_transit_minutes: int | None = Field(default=None, ge=1)
    pace: Pace = Pace.MODERATE
    day_start_local: str = DEFAULT_DAY_START_LOCAL
    day_end_local: str = DEFAULT_DAY_END_LOCAL
    lunch: MealWindow = MealWindow(
        start_local=DEFAULT_LUNCH_START,
        end_local=DEFAULT_LUNCH_END,
    )
    dinner: MealWindow = MealWindow(
        start_local=DEFAULT_DINNER_START,
        end_local=DEFAULT_DINNER_END,
    )
    transportation: TransportationPreference | None = None
    daily_budget: BudgetInput | None = None
    hidden_gems: bool = False
    group: GroupPlanning = Field(default_factory=GroupPlanning)
    show_transport_cost: bool = False
    accessibility: AccessibilityPreferences = Field(default_factory=AccessibilityPreferences)

    @model_validator(mode="after")
    def custom_transit(self) -> "TripControls":
        if self.max_transit_preset == MaxTransitPreset.CUSTOM and self.max_transit_minutes is None:
            raise ValueError("max_transit_minutes required when preset is custom")
        return self


class PlanRequest(ContractModel):
    user_prompt: str = Field(..., min_length=1)
    budget: BudgetInput | None = None
    transportation: TransportationPreference | None = None
    travel_dates: TravelDates
    location: LocationInput
    trip_controls: TripControls | None = None


class ExtractedPreferences(ContractModel):
    budget: BudgetInput | None = None
    dates: TravelDates | None = None
    interests: list[str] = Field(default_factory=list)
    pace: Pace = Pace.MODERATE
    food: FoodPreferences = Field(default_factory=FoodPreferences)
    hidden_gems: bool = False
    accessibility: AccessibilityPreferences = Field(default_factory=AccessibilityPreferences)
    group: GroupPlanning = Field(default_factory=GroupPlanning)
    transportation: TransportationPreference | None = None
    timezone: str | None = None

File: packages/contracts/typescript/src/preferences.ts
import { z } from "zod";

import {
  DEFAULT_DAY_END_LOCAL,
  DEFAULT_DAY_START_LOCAL,
  DEFAULT_DINNER_END,
  DEFAULT_DINNER_START,
  DEFAULT_LUNCH_END,
  DEFAULT_LUNCH_START,
  MEAL_MIN_DURATION_MINUTES,
} from "./constants";
import {
  BudgetTierSchema,
  GroupPrioritySchema,
  LocationTypeSchema,
  MaxTransitPresetSchema,
  OwnedVehicleSchema,
  PaceSchema,
  PublicTransportModeSchema,
} from "./enums";

export const BudgetInputSchema = z.object({
  daily_amount: z.number().positive(),
  currency: z.string().min(1),
  raw: z.string().optional().nullable(),
  tier: BudgetTierSchema.optional().nullable(),
});

export const TravelDatesSchema = z
  .object({
    start_date: z.string().date().optional().nullable(),
    end_date: z.string().date().optional().nullable(),
    duration_days: z.number().int().min(1).optional().nullable(),
  })
  .superRefine((value, ctx) => {
    const hasRange = Boolean(value.start_date && value.end_date);
    if (!hasRange && value.duration_days == null) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "Provide start_date+end_date or duration_days",
      });
    }
    if (hasRange && value.end_date! < value.start_date!) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "end_date must be on or after start_date",
      });
    }
  });

export const LocationInputSchema = z.object({
  query: z.string().min(1),
  type: LocationTypeSchema.optional().nullable(),
  radius_km: z.number().positive().optional().nullable(),
});

export const TransportationPreferenceSchema = z
  .object({
    owns_vehicle: z.boolean(),
    vehicle: OwnedVehicleSchema.optional().nullable(),
    public_mode: PublicTransportModeSchema.optional().nullable(),
  })
  .superRefine((value, ctx) => {
    if (value.owns_vehicle && !value.vehicle) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "vehicle is required when owns_vehicle is true",
      });
    }
  })
  .transform((value) => {
    if (!value.owns_vehicle && value.public_mode == null) {
      return { ...value, public_mode: "mixed" as const };
    }
    return value;
  });

export const TravelerSchema = z.object({
  name: z.string().optional().nullable(),
  interests: z.array(z.string()).default([]),
  is_team_lead: z.boolean().default(false),
});

export const GroupPlanningSchema = z.object({
  enabled: z.boolean().default(false),
  travelers: z.array(TravelerSchema).default([]),
  priority: GroupPrioritySchema.default("everyone"),
});

export const AccessibilityPreferencesSchema = z.object({
  kids: z.boolean().default(false),
  elderly: z.boolean().default(false),
  wheelchair: z.boolean().default(false),
  notes: z.string().optional().nullable(),
});

export const FoodPreferencesSchema = z.object({
  dietary_restrictions: z.array(z.string()).default([]),
  cuisines: z.array(z.string()).default([]),
});

export const MealWindowSchema = z.object({
  start_local: z.string(),
  end_local: z.string(),
  min_duration_minutes: z.number().int().min(60).default(MEAL_MIN_DURATION_MINUTES),
});

export const TripControlsSchema = z
  .object({
    max_transit_preset: MaxTransitPresetSchema.default("balanced"),
    max_transit_minutes: z.number().int().min(1).optional().nullable(),
    pace: PaceSchema.default("moderate"),
    day_start_local: z.string().default(DEFAULT_DAY_START_LOCAL),
    day_end_local: z.string().default(DEFAULT_DAY_END_LOCAL),
    lunch: MealWindowSchema.default({
      start_local: DEFAULT_LUNCH_START,
      end_local: DEFAULT_LUNCH_END,
      min_duration_minutes: MEAL_MIN_DURATION_MINUTES,
    }),
    dinner: MealWindowSchema.default({
      start_local: DEFAULT_DINNER_START,
      end_local: DEFAULT_DINNER_END,
      min_duration_minutes: MEAL_MIN_DURATION_MINUTES,
    }),
    transportation: TransportationPreferenceSchema.optional().nullable(),
    daily_budget: BudgetInputSchema.optional().nullable(),
    hidden_gems: z.boolean().default(false),
    group: GroupPlanningSchema.default({
      enabled: false,
      travelers: [],
      priority: "everyone",
    }),
    show_transport_cost: z.boolean().default(false),
    accessibility: AccessibilityPreferencesSchema.default({
      kids: false,
      elderly: false,
      wheelchair: false,
      notes: null,
    }),
  })
  .superRefine((value, ctx) => {
    if (value.max_transit_preset === "custom" && value.max_transit_minutes == null) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "max_transit_minutes required when preset is custom",
      });
    }
  });

export const PlanRequestSchema = z.object({
  user_prompt: z.string().min(1),
  budget: BudgetInputSchema.optional().nullable(),
  transportation: TransportationPreferenceSchema.optional().nullable(),
  travel_dates: TravelDatesSchema,
  location: LocationInputSchema,
  trip_controls: TripControlsSchema.optional().nullable(),
});

export const ExtractedPreferencesSchema = z.object({
  budget: BudgetInputSchema.optional().nullable(),
  dates: TravelDatesSchema.optional().nullable(),
  interests: z.array(z.string()).default([]),
  pace: PaceSchema.default("moderate"),
  food: FoodPreferencesSchema.default({
    dietary_restrictions: [],
    cuisines: [],
  }),
  hidden_gems: z.boolean().default(false),
  accessibility: AccessibilityPreferencesSchema.default({
    kids: false,
    elderly: false,
    wheelchair: false,
    notes: null,
  }),
  group: GroupPlanningSchema.default({
    enabled: false,
    travelers: [],
    priority: "everyone",
  }),
  transportation: TransportationPreferenceSchema.optional().nullable(),
  timezone: z.string().optional().nullable(),
});

export type BudgetInput = z.infer<typeof BudgetInputSchema>;
export type TravelDates = z.infer<typeof TravelDatesSchema>;
export type LocationInput = z.infer<typeof LocationInputSchema>;
export type TransportationPreference = z.infer<typeof TransportationPreferenceSchema>;
export type TripControls = z.infer<typeof TripControlsSchema>;
export type PlanRequest = z.infer<typeof PlanRequestSchema>;
export type ExtractedPreferences = z.infer<typeof ExtractedPreferencesSchema>;
